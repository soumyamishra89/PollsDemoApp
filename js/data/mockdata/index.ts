export const mockQuestions = [
    {
        "question": "Favorite programming language?",
        "published_at": "2020-04-15T14:49:59.810110+00:00",
        "url": "/questions/5",
        "choices": [
            {
                "choice": "Objective-C",
                "votes": 0,
                "url": "/questions/5/choices/19"
            },
            {
                "choice": "Python",
                "votes": 0,
                "url": "/questions/5/choices/18"
            },
            {
                "choice": "Ruby",
                "votes": 0,
                "url": "/questions/5/choices/20"
            },
            {
                "choice": "Swift",
                "votes": 0,
                "url": "/questions/5/choices/17"
            }
        ],
        "totalVotes": 0
    },
    {
        "question": "Capital of Germany?",
        "published_at": "2020-04-15T11:00:59.542382+00:00",
        "url": "/questions/4",
        "choices": [
            {
                "choice": "Berlin",
                "votes": 0,
                "url": "/questions/4/choices/13"
            },
            {
                "choice": "Frankfurt",
                "votes": 0,
                "url": "/questions/4/choices/16"
            },
            {
                "choice": "Hamburg",
                "votes": 0,
                "url": "/questions/4/choices/14"
            },
            {
                "choice": "Munich",
                "votes": 0,
                "url": "/questions/4/choices/15"
            }
        ],
        "totalVotes": 0
    },
    {
        "question": "test q?",
        "published_at": "2020-04-15T09:59:32.552020+00:00",
        "url": "/questions/3",
        "choices": [
            {
                "choice": "Objective-C",
                "votes": 0,
                "url": "/questions/3/choices/11"
            },
            {
                "choice": "Python",
                "votes": 0,
                "url": "/questions/3/choices/10"
            },
            {
                "choice": "Ruby",
                "votes": 0,
                "url": "/questions/3/choices/12"
            },
            {
                "choice": "Swift",
                "votes": 0,
                "url": "/questions/3/choices/9"
            }
        ],
        "totalVotes": 0
    },
    {
        "question": "",
        "published_at": "2020-04-10T16:05:25.127319+00:00",
        "url": "/questions/2",
        "choices": [
            {
                "choice": "Python",
                "votes": 14,
                "url": "/questions/2/choices/6"
            },
            {
                "choice": "Ruby",
                "votes": 6,
                "url": "/questions/2/choices/8"
            },
            {
                "choice": "Objective-C",
                "votes": 2,
                "url": "/questions/2/choices/7"
            },
            {
                "choice": "Swift",
                "votes": 1,
                "url": "/questions/2/choices/5"
            }
        ],
        "totalVotes": 23
    },
    {
        "question": "Favourite programming language?",
        "published_at": "2020-04-09T03:28:20.545996+00:00",
        "url": "/questions/1",
        "choices": [
            {
                "choice": "Python",
                "votes": 34,
                "url": "/questions/1/choices/2"
            },
            {
                "choice": "Objective-C",
                "votes": 8,
                "url": "/questions/1/choices/3"
            },
            {
                "choice": "Ruby",
                "votes": 4,
                "url": "/questions/1/choices/4"
            },
            {
                "choice": "Swift",
                "votes": 3,
                "url": "/questions/1/choices/1"
            }
        ],
        "totalVotes": 49
    }
]
