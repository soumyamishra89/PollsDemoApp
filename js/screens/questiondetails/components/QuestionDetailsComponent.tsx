import React, { useState } from 'react';
import { View, StyleSheet } from 'react-native';
import { useSafeArea } from 'react-native-safe-area-context';
import { Button } from 'react-native-paper';
import { PollQuestion, PollChoice } from '../../../data/types/PollsData';
import Styles from '../../../styles';
import Appbar from '../../common/compoonent/Appbar';
import ChoicesComponent from './ChoicesComponent';
import QuestionComponent from './QuestionComponent';
import colors from '../../../styles/colors';
import { NavigationProps } from '../../../data/types/ScreenProps';

interface Props extends NavigationProps {
    postVote: (choiceUrl: string) => Promise<boolean>,
    pollQuestion?: PollQuestion
}
export default function QuestionDetailsComponent(props: Props) {
    const [selectedChoice, setSelectedChoice] = useState<PollChoice | undefined>(undefined);
    const [loading, setLoading] = useState(false);

    const onVote = async () => {
        if (selectedChoice) {
            setLoading(true);
            const isVotingSuccess = await props.postVote(selectedChoice.url);
            if (isVotingSuccess) {
                props.navigation.goBack();
            } else {
                setLoading(false);
            }
        }
    }

    const inset = useSafeArea();
    return (
        <View style={[Styles.matchParent, {paddingBottom: inset.bottom}]}>
            <Appbar title="Poll Question"/>
            <View style={Styles.contentContainer}>
                <QuestionComponent question={props.pollQuestion?.question || ''} publishedDate={props.pollQuestion?.published_at.split('T')[0] || ''}/>
                <ChoicesComponent choices={props.pollQuestion?.choices || []} totalVotes={props.pollQuestion?.totalVotes || 0} selectedChoice={selectedChoice} onChoiceChange={setSelectedChoice}/>
            </View>
            <Button mode="contained" onPress={onVote} disabled={!selectedChoice} style={[Styles.button, !selectedChoice && Styles.disabledButton]} loading={loading}>Vote</Button>
        </View>
    )
}