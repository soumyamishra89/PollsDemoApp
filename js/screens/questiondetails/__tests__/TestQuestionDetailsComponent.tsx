import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import QuestionDetailsComponent from '../components/QuestionDetailsComponent';
import { mockQuestions } from '../../../data/mockdata';

describe("QuestionDetailsComponent", () => {
    describe("Snapshot test", () => {
        it("With question data", () => {            
            const pollQuestion = mockQuestions[0];
            const questionTree = renderer.create(<QuestionDetailsComponent postVote={jest.fn()}
            pollQuestion={pollQuestion} navigation={{}}/>).toJSON();
            expect(questionTree).toMatchSnapshot();

        });

    });
})