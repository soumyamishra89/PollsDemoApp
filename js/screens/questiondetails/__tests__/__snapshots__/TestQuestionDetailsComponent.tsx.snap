// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetailsComponent Snapshot test With question data 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "height": "100%",
        "width": "100%",
      },
      Object {
        "paddingBottom": 20,
      },
    ]
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Array [
        Object {
          "backgroundColor": "#3C7AFF",
          "elevation": 4,
          "zIndex": 0,
        },
        Object {
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 4,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3C7AFF",
          "elevation": 0,
          "flexDirection": "row",
          "height": 56,
          "marginTop": 0,
          "paddingHorizontal": 4,
        }
      }
    >
      <View
        style={
          Object {
            "width": 48,
          }
        }
      />
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "flex": 1,
              "paddingHorizontal": 12,
            },
            Array [
              false,
              Object {
                "alignItems": "center",
              },
              undefined,
            ],
          ]
        }
      >
        <Text
          accessibilityRole="header"
          accessibilityTraits="header"
          accessible={true}
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "#ffffff",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "fontSize": 17,
                },
                undefined,
              ],
            ]
          }
        >
          Poll Question
        </Text>
      </View>
      <View
        style={
          Object {
            "width": 48,
          }
        }
      />
    </View>
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "flex": 1,
        "marginTop": 16,
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 16,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "500",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                undefined,
              ],
              Array [
                Object {
                  "fontSize": 20,
                  "letterSpacing": 0.15,
                  "lineHeight": 30,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Favorite programming language?
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                undefined,
              ],
            ],
          ]
        }
      >
        2020-04-15
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 24,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#F1F5FF",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 24,
              "height": "100%",
              "position": "absolute",
            },
            false,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "borderColor": "#BFD3FE",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 20,
                "margin": 8,
                "width": 20,
              }
            }
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                false,
              ],
            ]
          }
        >
          Objective-C
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "marginRight": 8,
              },
              false,
            ],
          ]
        }
      >
        0.00%
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 24,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#F1F5FF",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 24,
              "height": "100%",
              "position": "absolute",
            },
            false,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "borderColor": "#BFD3FE",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 20,
                "margin": 8,
                "width": 20,
              }
            }
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                false,
              ],
            ]
          }
        >
          Python
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "marginRight": 8,
              },
              false,
            ],
          ]
        }
      >
        0.00%
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 24,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#F1F5FF",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 24,
              "height": "100%",
              "position": "absolute",
            },
            false,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "borderColor": "#BFD3FE",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 20,
                "margin": 8,
                "width": 20,
              }
            }
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                false,
              ],
            ]
          }
        >
          Ruby
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "marginRight": 8,
              },
              false,
            ],
          ]
        }
      >
        0.00%
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 24,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginVertical": 8,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#F1F5FF",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "borderRadius": 24,
              "height": "100%",
              "position": "absolute",
            },
            false,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <View
          accessibilityLiveRegion="polite"
          accessibilityRole="button"
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 18,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "borderColor": "#BFD3FE",
                "borderRadius": 10,
                "borderWidth": 2,
                "height": 20,
                "margin": 8,
                "width": 20,
              }
            }
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                false,
              ],
            ]
          }
        >
          Swift
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "marginRight": 8,
              },
              false,
            ],
          ]
        }
      >
        0.00%
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#D8D8D8",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "marginBottom": 16,
        "marginHorizontal": 16,
        "minWidth": 64,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.32)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Vote
        </Text>
      </View>
    </View>
  </View>
</View>
`;
